umap/umap_.py

```diff
@@ -677,152 +703,207 @@ def rdist(x, y):
 
-    for n in range(n_epochs):
-        for i in range(epochs_per_sample.shape[0]):
-            if epoch_of_next_sample[i] <= n:
-                j = head[i]
-                k = tail[i]
+        gamma: float (optional, default 1.0)
+            Weight to apply to negative samples.
 
-                current = head_embedding[j]
-                other = tail_embedding[k]
+        initial_alpha: float (optional, default 1.0)
+            Initial learning rate for the SGD.
 
-                dist_squared = rdist(current, other)
+        negative_sample_rate: int (optional, default 5)
+            Number of negative samples to use per positive sample.
 
-                grad_coeff = -2.0 * a * b * pow(dist_squared, b - 1.0)
-                grad_coeff /= a * pow(dist_squared, b) + 1.0
+        verbose: bool (optional, default False)
+            Whether to report information on the current progress of the algorithm.
 
-                for d in range(dim):
-                    grad_d = clip(grad_coeff * (current[d] - other[d]))
-                    current[d] += grad_d * alpha
-                    if move_other:
-                        other[d] += -grad_d * alpha
+        Returns
+        -------
+        embedding: array of shape (n_samples, n_components)
+            The optimized embedding.
+        """
 
-                epoch_of_next_sample[i] += epochs_per_sample[i]
+        dim = head_embedding.shape[1]
+        move_other = head_embedding.shape[0] == tail_embedding.shape[0]
+        alpha = initial_alpha
 
-                n_neg_samples = int(
-                    (n - epoch_of_next_negative_sample[i])
-                    / epochs_per_negative_sample[i]
-                )
+        epochs_per_negative_sample = epochs_per_sample / negative_sample_rate
+        epoch_of_next_negative_sample = epochs_per_negative_sample.copy()
+        epoch_of_next_sample = epochs_per_sample.copy()
 
-                for p in range(n_neg_samples):
-                    k = tau_rand_int(rng_state) % n_vertices
+        for n in range(n_epochs):
+            for i in range(epochs_per_sample.shape[0]):
+                if epoch_of_next_sample[i] <= n:
+                    j = head[i]
+                    k = tail[i]
+                    w_h = weight[i]
 
+                    current = head_embedding[j]
                     other = tail_embedding[k]
 
-                    dist_squared = rdist(current, other)
-
-                    grad_coeff = 2.0 * gamma * b
-                    grad_coeff /= (0.001 + dist_squared) * (
-                        a * pow(dist_squared, b) + 1
+                    dist_output, grad_dist_output = output_metric(
+                        current, other, *output_metric_kwds
                     )
 
-                    if not np.isfinite(grad_coeff):
-                        grad_coeff = 4.0
+                    w_l = pow((1 + a * pow(dist_output, 2 * b)), -1)
+                    if inverse:
+                        grad_coeff = (
```

Comment:
I admit this is not obvious to me -- can you sketch a derivation for this as the appropriate gradient? I would have thought that ``rhos[j]`` ought to also play a role?